src/Swift.Bindings/src/Emitter/StringEmitter/Handler/MethodHandler.cs

```diff
@@ -624,130 +691,164 @@ private void EmitSwiftError(IndentedTextWriter writer)
         {
             if (!_requiresIndirectResult)
             {
-                writer.WriteLine("this = result;");
+                csWriter.WriteLine("this = result;");
             }
         }
 
         /// <summary>
         /// Emits the return statement for the method.
         /// </summary>
-        /// <param name="writer">The IndentedTextWriter instance.</param>
-        private void EmitReturnMethod(IndentedTextWriter writer)
+        /// <param name="csWriter">The IndentedTextWriter instance.</param>
+        private void EmitReturnMethod(CSharpWriter csWriter)
         {
             if (_requiresIndirectResult)
             {
-                writer.WriteLine($"return SwiftMarshal.MarshalFromSwift<{_wrapperSignature.ReturnType}>((SwiftHandle)swiftIndirectResult.Value);");
+                csWriter.WriteLine($"return SwiftMarshal.MarshalFromSwift<{_wrapperSignature.ReturnType}>((SwiftHandle)swiftIndirectResult.Value);");
+                return;
             }
-            else
+
+            if (_requiresSwiftAsync)
             {
-                if (_env.MethodDecl.CSSignature.First().SwiftTypeSpec.IsEmptyTuple)
-                {
-                    writer.WriteLine("return;");
-                }
-                else
-                {
-                    writer.WriteLine("return result;");
-                }
+                csWriter.WriteLine("return task.Task;");
+                return;
             }
+
+            if (_env.MethodDecl.CSSignature.First().SwiftTypeSpec.IsEmptyTuple)
+            {
+                csWriter.WriteLine("return;");
+                return;
+            }
+
+            csWriter.WriteLine("return result;");
         }
 
         /// <summary>
         /// Emits the constructor signature.
         /// </summary>
-        /// <param name="writer">The IndentedTextWriter instance.</param>
-        private void EmitSignatureConstructor(IndentedTextWriter writer)
+        /// <param name="csWriter">The IndentedTextWriter instance.</param>
+        private void EmitSignatureConstructor(CSharpWriter csWriter)
         {
             var genericParams = _env.MethodDecl.IsGeneric switch
             {
                 true => $"<{string.Join(", ", _env.MethodDecl.GenericParameters.Select(p => _env.GenericTypeMapping[p.TypeName].PlaceholderName))}>",
                 false => ""
             };
-            writer.WriteLine($"public {_env.ParentDecl.Name}{genericParams}({_wrapperSignature.ParametersString()})");
+            csWriter.WriteLine($"public {_env.ParentDecl.Name}{genericParams}({_wrapperSignature.ParametersString()})");
         }
 
         /// <summary>
         /// Emits the method signature.
         /// </summary>
         /// <param name="writer">The IndentedTextWriter instance.</param>
-        private void EmitSignatureMethod(IndentedTextWriter writer)
+        private void EmitSignatureMethod(CSharpWriter csWriter)
         {
+            if (_requiresSwiftAsync)
```

Comment:
is there a reason this lives here and not in a separate method. I am not sure how it is related to `Signature` emission

Reply:
Moved to a separate method.